habitat_baselines/il/data/edfe_data.py

```diff
@@ -0,0 +1,224 @@
+from tqdm import tqdm
+
+import habitat
+from habitat.core.simulator import ShortestPathPoint
+
+
+class EDFEDataset(Dataset):
+    """Pytorch dataset for Embodied Q&A's feature-extractor"""
+
+    def __init__(self, env, config, mode="train"):
+        """
+        Args:
+            env (habitat.Env): Habitat environment
+            config: Config
+            mode: 'train'/'val'
+        """
+        self.env = env
+        self.sim = self.env.sim
+        self.config = config.TASK_CONFIG
+        self.mode = mode
+
+        self.episodes = self.get_all_episodes(config)
+
+        self.dataset_path = config.DATASET_PATH
+
+        self.scene_ids = []
+        self.scene_episode_dict = {}
+
+        # preparing a dict that stores list of episodes for each scene
+        for ep in self.episodes:
+            if ep.scene_id not in self.scene_ids:
+                self.scene_ids.append(ep.scene_id)
+                self.scene_episode_dict[ep.scene_id] = [ep.episode_id]
+            else:
+                self.scene_episode_dict[ep.scene_id].append(ep.episode_id)
+
+        self.disk_cache_exists = self.check_cache_exists()
+
+        if not self.disk_cache_exists:
+            try:
+                self.make_dataset_dirs()
+            except FileExistsError:
+                pass
+            """
+            for each scene > load scene in memory > save frames for each
+            episode corresponding to each scene
+            """
+            print("Saving rgb, seg, depth frames to disk.")
+
+            for scene in tqdm(list(self.scene_episode_dict.keys())):
+
+                self.load_scene(scene)
+
+                for ep_id in tqdm(self.scene_episode_dict[scene]):
+
+                    episode = next(
+                        ep for ep in self.episodes if ep.episode_id == ep_id
+                    )
+
+                    pos_queue = episode.shortest_paths[0]
+                    random_pos = random.sample(pos_queue, 9)
+                    self.save_frames(random_pos, ep_id)
+
+            print("Saved all episodes' frames to disk. Frame dataset ready.")
+        self.env.close()
+
+        self.rgb_list = sorted(
+            os.listdir(self.dataset_path.format(split=self.mode, type="rgb"))
+        )
+        self.depth_list = sorted(
+            os.listdir(self.dataset_path.format(split=self.mode, type="depth"))
+        )
+        self.seg_list = sorted(
+            os.listdir(self.dataset_path.format(split=self.mode, type="seg"))
+        )
+
+    def save_frames(
+        self, pos_queue: List[ShortestPathPoint], episode_id
+    ) -> None:
+        r"""Writes rgb, seg, depth frames to disk.
+        """
+
+        for idx, pos in enumerate(pos_queue):
+
+            observation = self.env.sim.get_observations_at(
+                pos.position, pos.rotation
+            )
+
+            depth = observation["depth"]
+            rgb = observation["rgb"]
+            seg = observation["semantic"]
+            seg = seg % 40
```

Comment:
This likely isn't correct.  The semantic sensor produces instances segmentation.  If you want semantic segmentation, you'll need to map object ID's to semantic label ID.

Reply:
Thank you @erikwijmans for pointing this out.

With help from [this issue](https://github.com/facebookresearch/habitat-sim/issues/263), I am able to extract the semantic label ids using this -

```
scene = self.env.sim.semantic_annotations()

instance_id_to_label_id = {int(obj.id.split("_")[-1]): obj.category.index() for obj in scene.objects}
self.mapping = np.array([ instance_id_to_label_id[i] for i in range(len(instance_id_to_label_id)) ])
seg = np.take(self.mapping, observation['semantic'])
```

I see that there are a total of 42 categories (semantic label ids) across all the objects in the MP3D scene dataset-
|Category ID   | Category name   |
|:-:|:-:|
| -1  | '' |
| 0  | 'void'  |
| 1  | 'wall'  |
| ...  |  ... |
| 40  | 'misc'  |

Just wanted to confirm if it'd be okay to fuse the first two categories (ids -1 and 0) into one and specify 41 classes for the task of learning the segmentation map?

Reply:
Yeah, that sounds fine.